Bluetooth: controller: Ignore connections from same peer

wopu-ot opened the pull request:
Ignore connection indications from peers that are already
connected. This is to bring the behavior of the controller in
accordance with [5.2, Vol 6, Part B, 4.5 Connection state]:
"If an advertiser receives a connection request from an initiator it
is already connected to, it shall ignore that request."

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

Fixes #26172

wopu-ot committed:
Bluetooth: controller: Ignore connections from same peer

Ignore connection indications from peers that are already
connected. This is to bring the behavior of the controller in
accordance with [5.2, Vol 6, Part B, 4.5 Connection state]:
"If an advertiser receives a connection request from an initiator it
is already connected to, it shall ignore that request."

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -77,6 +78,46 @@ void ull_slave_setup(memq_link_t *link, struct node_rx_hdr *rx,
+
+#if defined(CONFIG_BT_CTLR_IGNORE_SAME_PEER_CONN)
+	uint8_t own_addr_type = pdu_adv->rx_addr;
+	uint8_t *own_addr = ll_addr_get(own_addr_type, NULL);
```
Why not use the `pdu_adv->connect_ind.adv_addr`?

wopu-ot replied:
The intention is to ignore connections to the same identity; `pdu_adv->connect_ind.adv_addr` may be an RPA that is refreshed periodically without changing identity.

cvinayak replied:
My concern is regarding extended advertising where each advertising set has its own random address.

wopu-ot replied:
Good point. I will add a field to `ll_adv_set` to avoid races between different advertising sets.

wopu-ot committed:
Bluetooth: controller: Rename peer's address in advertising set struct

Rename the peer's address in the advertising set struct from `id_addr`
to `peer_addr` to clarify what the address refers to.

Signed-off-by: Wolfgang Puffitsch <wopu@demant.com>

Thalley commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -372,6 +372,15 @@ config BT_CTLR_CONN_RSSI
 	help
 	  Enable connection RSSI measurement.
 
+config BT_CTLR_IGNORE_SAME_PEER_CONN
+	bool "Ignore connection requests from same peer"
+	default y
+	help
+	  Ignored connection requests from the same peer. While the
+	  Bluetooth specification does not allow multiple connections
+	  with the same peer, allowing such connections is useful
+	  while debugging multiple connections.
```
If the spec specifies a `shall` for this, why is it a Kconfig? Wouldn't disabling this cause the stack to be non-compliant?

wopu-ot replied:
This was suggested by @carlescufi https://github.com/zephyrproject-rtos/zephyr/issues/26172#issuecomment-644093770

Thalley replied:
I see. Then I would suggest only allowing it to be turned on for TEST/DEBUG builds, e.g. by only allowing it be be `n` if e.g. `BT_TESTING` is enabled. 

I don't think it's a good idea to intentionally allowing a non-spec compliant build without some sort of warning or making sure that whoever disables this is aware of the consequences.

wopu-ot replied:
I could make it dependent on `BT_CTLR_ADVANCED_FEATURES`, though I believe I would have to flip the polarity (`...ALLOW...` instead of `...IGNORE...` and `!defined` instead of `defined`) to get the correct behavior if the former is not defined.

Thalley replied:
Alternatively we could add similar warnings to disabling this as we have when e.g. we enable SMP debug keys. See https://github.com/zephyrproject-rtos/zephyr/pull/26186/files

wopu-ot replied:
I moved the user-controller flag it under `BT_CTLR_ADVANCED_FEATURES` and added a WARNING to the description. I also made the configuration now dependent on having more than one connection to not punish implementations that allow only a single connection anyway.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_filter.c:
```diff
@@ -1081,6 +1091,28 @@ static void filter_clear(struct lll_filter *filter)
+		if (conn &&
+		    conn->peer_addr_type < 2U &&
+		    !memcmp(conn->peer_addr, rl[rl_idx].curr_rpa.val, BDADDR_SIZE)) {
+			memcpy(conn->peer_addr, rl[rl_idx].id_addr.val,
+			       BDADDR_SIZE);
```
Minor: The maximum line length was recently increased to 100, so this can be on a single line

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -889,6 +889,13 @@ uint8_t ll_adv_enable(uint8_t enable)
 
+#if defined(CONFIG_BT_CTLR_IGNORE_SAME_PEER_CONN)
+		conn->own_addr_type = 0;
+		memset(conn->own_addr, 0xFF, sizeof(conn->own_addr));
+		conn->peer_addr_type = 0;
+		memset(conn->peer_addr, 0xFF, sizeof(conn->peer_addr));
```
What's the reasoning behind setting the addresses to 0xFFFFFF?

wopu-ot replied:
0xFFFFFF matches `BT_ADDR_NONE`, which I find slightly more intuitive than 0x000000, which would match `BT_ADDR_ANY`.

Thalley replied:
I see. In the case that a literal value (in this case 0xFFFFFF) has a specific meaning (e.g. to be comparable with BT_ADDR_NONE), then I strongly suggest that we add a `#define` or similar for the value.

wopu-ot replied:
Copying from `BT_ADDR_LE_NONE` now to make the meaning clearer.

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -475,6 +475,18 @@ config BT_CTLR_CONN_RSSI_EVENT
+	  with the same peer, allowing such connections is useful
+	  while debugging multiple connections.
+
+	  WARNING: This option enables behavior that violates the Bluetooth
+	  specification.
```
I would like @carlescufi input on whether this is enough, or if we also should add build warnings like was done for other such configurations in https://github.com/zephyrproject-rtos/zephyr/pull/26186

wopu-ot replied:
> I would like @carlescufi input on whether this is enough, or if we also should add build warnings like was done for other such configurations in #26186

@carlescufi What is your take in this?

Thalley replied:
@wopu-ot Is this something that is critical to get in? I can mark it as approved, and then we can consider adding additional warnings (if needed) afterwards in a separate PR.

erbr-ot reviewed (approved):
LGTM

cvinayak reviewed (approved):
quick-ack.
